pallets/foreign-investments/src/types.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge Chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_types::investments::Swap;
+use codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::traits::{EnsureAdd, EnsureSub};
+
+/// Reflects the reason for the last token swap update such that it can be
+/// updated accordingly if the last and current reason mismatch.
+#[derive(
+	Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+pub enum TokenSwapReason {
+	Investment,
+	Redemption,
+}
+
+/// Reflects all states a foreign investment can have until it is processed as
+/// an investment via `<T as Config>::Investment`. This includes swapping it
+/// into a pool currency or back, if the investment is decreased before it is
+/// fully processed.
+#[derive(
+	Clone, Default, PartialOrd, Ord, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+pub enum InvestState<
```

Comment:
Yeah, the tight deadline doesn't help.

> the inner states should be bounded vectors which adds a lot of boilerplate on its own.

If the maximum states are bounded and states are uniques in that vector, we could create a 3-optional-tuple type instead to use a `BoundedVec`. Or maybe a wrapper over a `BoundedSet` with some function to handle it easily.

Probably it requires some time to stop and think how to do it well, so maybe it's better to do it in the future if deserve the efforts.

> Going with the below approach, there would be inner state combinations which must never occur. The current approach is more intuitive when looking at the state diagrams.

I think this would not be a problem, the logic will be agnostic of which combinations would exist or not, isn't it? The logic should only process each state found in the vector independently if the combination of those makes sense, I guess

Reply:
> I think this would not be a problem, the logic will be agnostic of which combinations would exist or not, isn't it? The logic should only process each state found in the vector independently if the combination of those makes sense, I guess.

Yes, you are right. 

> Probably it requires some time to stop and think how to do it well, so maybe it's better to do it in the future if deserve the efforts.

To me it definitely feels out of scope, especially because it does not provide any benefits apart from potentially better readability and maintenance. If we are super lucky and testing goes well, we might have capacity to look into that before rolling out to production 🙃

Reply:
Sure! No hurry / required at all!

Reply:
Just dropping some thoughts regarding this topic (not for now). 

I think `InvestState` could be represented as a *product type* as follows (no longer needed `BoundedSet`):

```rust
struct InvestState {
    pool_swap: Option<Swap<Balance, Currency>>,
    foreign_swap: Option<Swap<Balance, Currency>>,
    done_amount: Balance, // If zero means no state `SwapIntoForeignDone`
    invest_amount: Balance, // If zero means no `InvestmentOngoing`
}
```

The same will happen for `ReedemState`:
```rust
struct ReedemState {
    invested_amount: Balance,
    reedem_amount: Balance,
    collectable_redemption: bool,
    foreign_swap: Option<Swap<Balance, Currency>>,
    done_amount: Balance,
}
```

Later the transition methods will check the fields to know how to change that state.